unison-src/transcripts/fix2822.output.md

```diff
@@ -0,0 +1,141 @@
+# Inability to reference a term or type with a name that has segments starting with an underscore
+
+There should be no issue having terms with an underscore-led component
+
+``` unison
+_a.blah = 2
+
+b = _a.blah + 1
+```
+
+``` ucm
+
+  Loading changes detected in scratch.u.
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      _a.blah : Nat
+      b       : Nat
+
+```
+Or even that *are* a single “blank” component
+
+``` unison
+_b = 2
+
+x = _b + 1
+```
+
+``` ucm
+
+  Loading changes detected in scratch.u.
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      _b : Nat
+      x  : Nat
+
+```
+Types can also have underscore-led components.
+
+``` unison
+unique type _a.Blah = A
+
+c : _a.Blah
+c = A
+```
+
+``` ucm
+
+  Loading changes detected in scratch.u.
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      type _a.Blah
+      c : Blah
+
+```
+And we should also be able to access underscore-led fields.
+
+``` unison
+type Hello = {_value : Nat}
+
+doStuff = _value.modify
+```
+
+``` ucm
+
+  Loading changes detected in scratch.u.
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      type Hello
+      Hello._value        : Hello -> Nat
+      Hello._value.modify : (Nat ->{g} Nat) -> Hello ->{g} Hello
+      Hello._value.set    : Nat -> Hello -> Hello
+      doStuff             : (Nat ->{g} Nat) -> Hello ->{g} Hello
+
+```
+But pattern matching shouldn’t bind to underscore-led names.
+
+``` unison
+dontMap f = cases
+  None -> false
+  Some _used -> f _used
+```
+
+``` ucm
+
+  Loading changes detected in scratch.u.
+
+  I couldn't figure out what _used refers to here:
+  
+      3 |   Some _used -> f _used
+  
```

Comment:
It would be nice if we could give a hint here, when an unknown identifier matches an unbound pattern var, but it doesn't have to be in this PR.

Reply:
Yeah, agreed. I don’t think it should be in this PR, because it’s not a parser issue, and this behavior already existed.

And this is another case where using color isn’t enough to convey the error. It should be more like
```
      3 |   Some _used -> f _used
                            ^^^^^
```
(So, another vote for integrating [Diagnose](https://github.com/mesabloo/diagnose#readme).)

Reply:
> (So, another vote for integrating [Diagnose](https://github.com/mesabloo/diagnose#readme).)

Whoa that looks crazy. I'm interested.

Reply:
Long ago we had started implementing something like what you showed, but we weren't good at it and we scrapped it. But off-the-shelf sounds nice.

Reply:
To give credit where it‘s due, @ChrisPenner is the one who shared it initially (https://discord.com/channels/862108724948500490/1200119435931942994/1261743434704879796).